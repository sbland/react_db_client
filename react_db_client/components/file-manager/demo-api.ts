import { EFileType, IFile } from '@react_db_client/constants.client-types';

export const asyncFileUpload = async (
  data: File,
  fileType: EFileType,
  callback: () => void,
  metaData: Partial<IFile>
) => {};
