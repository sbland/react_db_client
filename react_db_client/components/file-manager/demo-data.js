export const demoSearchResults = [
    { uid: 'abc', name: 'abc.jpg', fileType: 'image' },
    { uid: 'foo', name: 'foo.jpg', fileType: 'image' },
    { uid: 'bar', name: 'bar.jpg', fileType: 'image' },
  ]