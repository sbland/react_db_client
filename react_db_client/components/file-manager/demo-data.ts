import { EFileType, IFile } from '@react_db_client/constants.client-types';

export const demoSearchResults: IFile[] = [
  {
    uid: 'abc',
    name: 'abc.jpg',
    label: 'abc',
    fileType: EFileType.IMAGE,
    filePath: 'files/abc.jpg',
  },
  {
    uid: 'foo',
    name: 'foo.jpg',
    label: 'foo',
    fileType: EFileType.IMAGE,
    filePath: 'files/foo.jpg',
  },
  {
    uid: 'bar',
    name: 'bar.jpg',
    label: 'bar',
    fileType: EFileType.IMAGE,
    filePath: 'files/bar.jpg',
  },
];
