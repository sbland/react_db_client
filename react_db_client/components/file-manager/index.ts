export { FileManager } from './file-manager';
export type { IFileManagerProps } from './file-manager';