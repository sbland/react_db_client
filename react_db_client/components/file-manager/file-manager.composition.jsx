import React from 'react';
import { CompositionWrapDefault } from '@react_db_client/helpers.composition-wraps';
import { FileManager } from './file-manager';
import { demoSearchResults } from './demo-data';



const defaultProps = {
  handleSelect: () => alert('Selected'),
  collectionId: 'projects',
  documentId: 'Example Project',
  fileType: 'image',
  allowMultiple: false,
  asyncGetDocuments: async() => demoSearchResults,
  fileServerUrl: 'fileserverurl',
};

export const BasicFileManager = () => (
  <CompositionWrapDefault height="4rem" width="8rem">
    <FileManager {...defaultProps} />
  </CompositionWrapDefault>
);
