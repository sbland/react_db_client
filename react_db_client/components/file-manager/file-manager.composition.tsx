import React from 'react';
import { screen } from '@testing-library/react';
import { EFileType, IFile } from '@react_db_client/constants.client-types';
import { CompositionWrapDefault } from '@react_db_client/helpers.composition-wraps';
import { FileManager, IFileManagerProps } from './file-manager';
import { demoSearchResults } from './demo-data';
import { asyncFileUpload } from './demo-api';

const defaultProps: IFileManagerProps = {
  handleSelect: (args: any) => alert('Selected'),
  fileType: EFileType.IMAGE,
  allowMultiple: false,
  asyncGetFiles: async () => demoSearchResults,
  fileServerUrl: 'fileserverurl',
  asyncFileUpload,
};

export const BasicFileManager = () => {
  const [savedFiles, setSavedFiles] = React.useState(demoSearchResults);
  const [selection, setSelection] = React.useState<IFile | null>(null);
  const handleFileUpload = async (
    data: File,
    fileType: EFileType,
    callback: () => void,
    metaData: Partial<IFile>
  ) => {
    asyncFileUpload(data, fileType, callback, metaData);
    const fileMetaData: IFile = {
      ...metaData,
      uid: data.name,
      label: data.name,
      name: data.name,
      filePath: '',
      fileType,
      data,
    };
    setSavedFiles((prev) => [...prev, fileMetaData]);
  };
  return (
    <>
      <CompositionWrapDefault height="30rem" width="40rem">
        <FileManager
          {...defaultProps}
          asyncFileUpload={handleFileUpload}
          asyncGetFiles={async () => savedFiles}
          handleSelect={(f) => setSelection(f as IFile)}
        />
      </CompositionWrapDefault>
      <p data-testid="curSel">{selection?.uid}</p>
    </>
  );
};

BasicFileManager.waitForReady = async () => {
  await screen.findByText(demoSearchResults[0].name);
};

export const BasicFileManagerMultiple = () => {
  const [savedFiles, setSavedFiles] = React.useState(demoSearchResults);
  const [selection, setSelection] = React.useState<IFile[]>([]);
  const handleFileUpload = async (
    data: File,
    fileType: EFileType,
    callback: () => void,
    metaData: Partial<IFile>
  ) => {
    asyncFileUpload(data, fileType, callback, metaData);
    const fileMetaData: IFile = {
      uid: data.name,
      label: data.name,
      name: data.name,
      filePath: '',
      fileType,
      data,
    };
    setSavedFiles((prev) => [...prev, fileMetaData]);
  };

  const handleSelect = (f) => {
    setSelection((prev) => [...prev, ...f]);
  };

  return (
    <>
      <CompositionWrapDefault height="30rem" width="40rem">
        <FileManager
          {...defaultProps}
          asyncFileUpload={handleFileUpload}
          asyncGetFiles={async () => savedFiles}
          handleSelect={handleSelect}
          allowMultiple
        />
      </CompositionWrapDefault>
      <p data-testid="curSel">{selection.map((s) => s.uid).join(',')}</p>
    </>
  );
};

BasicFileManagerMultiple.waitForReady = async () => {
  await screen.findByText(demoSearchResults[0].name);
};
