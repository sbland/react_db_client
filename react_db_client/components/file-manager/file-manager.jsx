import PropTypes from 'prop-types';
import React, { useState } from 'react';
import { FileUploader } from '@react_db_client/components.file-uploader';
import { SearchAndSelect } from '@react_db_client/components.search-and-select';
import { searchFilesFunction, searchResultHeadings } from './logic';

export const FileManager = ({
  handleSelect,
  collectionId,
  documentId,
  fileType,
  allowMultiple,
  asyncGetDocuments,
  fileServerUrl,
}) => {
  const [forceUpdate, setForceUpdate] = useState(0);
  return (
    <div className="fileManager_wrap sectionWrapper">
      <section>
        <h2>Select File</h2>
        <br />
        <br />
        {/* TODO: Should refresh on file upload */}
        <SearchAndSelect
          searchFunction={searchFilesFunction(asyncGetDocuments)(collectionId, documentId, fileType)}
          initialFilters={[]}
          handleSelect={handleSelect}
          autoUpdate
          allowFilters={false}
          headings={searchResultHeadings(fileServerUrl, collectionId, documentId)}
          showSearchField
          // searchFieldTargetField="name"
          key={forceUpdate}
          returnFieldOnSelect="filePath"
          allowMultiple={allowMultiple}
          showRefreshBtn
        />
      </section>
      <section>
        <h2>Upload New File</h2>
        <FileUploader
          collectionId={collectionId}
          documentId={documentId}
          fileType={fileType}
          onUpload={() => setForceUpdate(forceUpdate + 1)}
        />
      </section>
    </div>
  );
};

FileManager.propTypes = {
  handleSelect: PropTypes.func.isRequired,
  collectionId: PropTypes.string.isRequired,
  documentId: PropTypes.string.isRequired,
  fileType: PropTypes.oneOf(['image', 'document', 'data', '*']),
  allowMultiple: PropTypes.bool,
  asyncGetDocuments: PropTypes.func.isRequired,
  fileServerUrl: PropTypes.string.isRequired,
};

FileManager.defaultProps = {
  allowMultiple: false,
  fileType: '*',
};

