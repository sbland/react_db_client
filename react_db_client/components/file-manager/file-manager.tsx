import PropTypes from 'prop-types';
import React, { useState } from 'react';
import { EFileType, IFile } from '@react_db_client/constants.client-types';
import { FileUploader } from '@react_db_client/components.file-uploader';
import {
  SearchAndSelect,
  TSearchAndSelectSearchFunction,
} from '@react_db_client/components.search-and-select-v2';
import { searchResultHeadings, searchResultImageHeadings } from './logic';

export interface IFileManagerProps {
  handleSelect: (fileData: null | IFile | IFile[]) => void;
  collectionId?: never;
  documentId?: never;
  fileType: EFileType;
  allowMultiple?: boolean;
  asyncGetFiles: TSearchAndSelectSearchFunction<IFile>;
  fileServerUrl: string;
  asyncFileUpload: (
    data: File,
    fileType: EFileType,
    callback: () => void,
    metaData: Partial<IFile>
  ) => Promise<void>;
}

export const availableFilters = {}; // TODO: Setup available file filters

export const FileManager: React.FC<IFileManagerProps> = ({
  handleSelect,
  collectionId,
  documentId,
  fileType,
  allowMultiple,
  asyncGetFiles,
  fileServerUrl,
  asyncFileUpload,
}) => {
  const [forceUpdate, setForceUpdate] = useState(0);
  if (collectionId || documentId)
    throw new Error('No longer need collection id and doc id');

  return (
    <div
      className="fileManager_wrap sectionWrapper"
      data-testid="rdc-file-manager"
    >
      <section>
        <h2>Select File</h2>
        <br />
        {/* TODO: Should refresh on file upload */}
        <SearchAndSelect<IFile>
          id="file-manager-existing-files"
          searchFunction={asyncGetFiles}
          initialFilters={[]}
          handleSelect={handleSelect}
          autoUpdate
          allowFilters={false}
          availableFilters={availableFilters}
          headings={
            fileType === EFileType.IMAGE
              ? searchResultImageHeadings(fileServerUrl)
              : searchResultHeadings(fileServerUrl)
          }
          showSearchField
          // searchFieldTargetField="name"
          key={forceUpdate}
          // returnFieldOnSelect="filePath"
          allowMultiple={allowMultiple}
          showRefreshBtn
        />
      </section>
      <section>
        <h2>Upload New File</h2>
        <FileUploader
          fileType={fileType}
          onUpload={() => setForceUpdate(forceUpdate + 1)}
          asyncFileUpload={asyncFileUpload}
        />
      </section>
    </div>
  );
};

FileManager.propTypes = {
  handleSelect: PropTypes.func.isRequired,
  fileType: PropTypes.oneOf(['image', 'document', 'data', '*']),
  allowMultiple: PropTypes.bool,
  asyncGetFiles: PropTypes.func.isRequired,
  fileServerUrl: PropTypes.string.isRequired,
};

FileManager.defaultProps = {
  allowMultiple: false,
  fileType: EFileType.ANY,
};
