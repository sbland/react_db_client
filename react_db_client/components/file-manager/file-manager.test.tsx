import '@samnbuk/react_db_client.testing.enzyme-setup';
import React from 'react';
import { mount, shallow } from 'enzyme';
import { MockReactC } from '@react_db_client/testing.utils';
import { SearchAndSelect } from '@react_db_client/components.search-and-select';
import { EFileType } from '@react_db_client/constants.client-types';

import { FileManager } from './file-manager';
import * as compositions from './file-manager.composition';
import { demoSearchResults } from './demo-data';

jest.mock('@react_db_client/components.search-and-select', () =>
  MockReactC('SearchAndSelect', ['SearchAndSelect'])
);
jest.mock('@react_db_client/components.file-uploader', () =>
  MockReactC('FileUploader', ['FileUploader'])
);

Date.now = jest.fn(() => 123); //14.02.2017
const handleSelect = jest.fn();
const asyncFileUpload = jest.fn();
const asyncGetFiles = jest
  .fn()
  .mockReturnValue(new Promise((resolve) => resolve(demoSearchResults)));

const defaultProps = {
  handleSelect,
  fileType: EFileType.IMAGE,
  allowMultiple: false,
  asyncGetFiles,
  fileServerUrl: 'fileserverurl',
  asyncFileUpload,
};

describe('file-manager', () => {
  beforeEach(() => {
    handleSelect.mockClear();
    asyncGetFiles.mockClear();
  });
  test('Renders', () => {
    shallow(<FileManager {...defaultProps} />);
  });

  describe('Compositions', () => {
    Object.entries(compositions).forEach(([name, Composition]) => {
      test(name, () => {
        mount(<Composition />);
      });
    });
  });
  describe('shallow renders', () => {
    test('Matches Snapshot', () => {
      const component = shallow(<FileManager {...defaultProps} />);
      const tree = component.debug();
      expect(tree).toMatchSnapshot();
    });

    describe('Unit Testing', () => {
      let component;
      beforeEach(() => {
        component = mount(<FileManager {...defaultProps} />);
      });
      test('should call asyncGetDocuments when searchFunction called', () => {
        const sortBy = 'name';
        const searchString = 'searchStr';
        const filters = [];
        const schema = ['uid', 'name', 'updatedAt', 'createdAt', 'fileType', 'filePath'];

        const sas = component.find(SearchAndSelect);
        sas.props().searchFunction([], sortBy, searchString, schema);

        expect(asyncGetFiles).toHaveBeenCalledWith(filters, sortBy, searchString, schema);
      });
      test('should call handle select when search and select handle select called', () => {
        const sas = component.find(SearchAndSelect);
        const selectionData = {
          uid: 'abc',
          name: 'ABC',
        };
        sas.props().handleSelect(selectionData.uid, selectionData);
        expect(handleSelect).toHaveBeenCalledWith(selectionData.uid, selectionData);
      });
    });
  });
});
