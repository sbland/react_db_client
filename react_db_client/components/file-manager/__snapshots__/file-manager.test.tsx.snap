// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`file-manager shallow renders Matches Snapshot 1`] = `
"<div className=\\"fileManager_wrap sectionWrapper\\">
  <section>
    <h2>
      Select File
    </h2>
    <br />
    <br />
    <SearchAndSelect searchFunction={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} initialFilters={{...}} handleSelect={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} autoUpdate={true} allowFilters={false} availableFilters={{...}} headings={{...}} showSearchField={true} returnFieldOnSelect=\\"filePath\\" allowMultiple={false} showRefreshBtn={true} />
  </section>
  <section>
    <h2>
      Upload New File
    </h2>
    <FileUploader fileType=\\"image\\" onUpload={[Function: onUpload]} asyncFileUpload={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} />
  </section>
</div>"
`;
